critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -0,0 +1,256 @@
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.critwall;
+
+import ca.nrc.cadc.auth.AuthMethod;
+import ca.nrc.cadc.io.ReadException;
+import ca.nrc.cadc.io.WriteException;
+import ca.nrc.cadc.net.FileContent;
+import ca.nrc.cadc.net.HttpGet;
+import ca.nrc.cadc.net.HttpPost;
+import ca.nrc.cadc.net.ResourceAlreadyExistsException;
+import ca.nrc.cadc.net.ResourceNotFoundException;
+import ca.nrc.cadc.net.TransientException;
+import ca.nrc.cadc.reg.Standards;
+import ca.nrc.cadc.reg.client.RegistryClient;
+import ca.nrc.cadc.vos.Direction;
+import ca.nrc.cadc.vos.Protocol;
+import ca.nrc.cadc.vos.Transfer;
+import ca.nrc.cadc.vos.TransferParsingException;
+import ca.nrc.cadc.vos.TransferReader;
+import ca.nrc.cadc.vos.TransferWriter;
+import ca.nrc.cadc.vos.VOS;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.Artifact;
+import org.opencadc.inventory.InventoryUtil;
+import org.opencadc.inventory.db.ArtifactDAO;
+import org.opencadc.inventory.storage.NewArtifact;
+import org.opencadc.inventory.storage.StorageAdapter;
+
+import org.opencadc.inventory.storage.StorageEngageException;
+import org.opencadc.inventory.storage.StorageMetadata;
+
+public class FileSyncJob implements Runnable {
+    private static final Logger log = Logger.getLogger(FileSyncJob.class);
+
+    private final ArtifactDAO artifactDAO;
+    private final URI artifactID;
+    private final URI resourceID;
+    private final StorageAdapter storageAdapter;
+
+    public FileSyncJob(URI artifactID, URI resourceID, StorageAdapter storageAdapter, ArtifactDAO artifactDAO) {
+        InventoryUtil.assertNotNull(FileSyncJob.class, "artifactID", artifactID);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "resourceID", resourceID);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "storageAdapter", storageAdapter);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "artifactDAO", artifactDAO);
+
+        this.artifactID = artifactID;
+        this.resourceID = resourceID;
+        this.storageAdapter = storageAdapter;
+        this.artifactDAO = artifactDAO;
+    }
+
+    @Override
+    public void run() {
+
+        final List<URL> urlList;
+        try {
+            urlList = getdownloadURLs();
+        } catch (Exception e) {
+            // fail - nothing can be done without the list, negotiation cannot be retried
+            log.error("transfer negotiation failed.");
+            return;
+        }
+        log.debug("endpoints returned: " + urlList);
+
+        int retryCount = 0;
+        Iterator<URL> urlIterator = urlList.iterator();
+
```

Comment:
with the changes in syncArtifact, this `while` loop needs to be wrapped in a try/catch for then fatal errors. The catch can be more or less like the one for failed negoitiaton (log and return -- and I do like the return in there so it is explicit that nothing more can be done even when it is near or at theb end of run()). 

Since the HttpGet.prepare() inside syncArtifact does some retrying already (mainly 503s from server), the while loop here is mainly compensating for failed in transit (ReadException during the put)... with any retry loop you need a delay (Thread.sleep(milliseconds) which should be down by the `retryCount++`... a compact way to do it with increasing delay is declare a constant like:
```
static long[] RETRY_DELAY = new long[] { 2000L, 4000L };
```
then use `while (retryCount < RETRY_DELAY.length)` and `Thread.sleep(RETRY_DELAY[retryCount++]);` and increasing the number of retries means adding delays to the array.

Reply:
done